File: 20190519python/0687. Longest Univalue Path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        if not root:
            return 0
        return max(self.longestUnivaluePath(root.left),self.longestUnivaluePath(root.right),self.theSameValPath(root.left,root.val)+self.theSameValPath(root.right,root.val))
    
    def theSameValPath(self,root,target):
        if not root or root.val!=target:
            return 0
        return 1+max(self.theSameValPath(root.left,target),self.theSameValPath(root.right,target))
        

File: archives/C#/0687. Longest Univalue Path.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int LongestUnivaluePath(TreeNode root) {
        if(root==null){
            return 0;
        }
        int maxLongestUnivaluePath=Math.Max(LongestUnivaluePath(root.left),LongestUnivaluePath(root.right));
        int maxSumLongestPath=GetmaxLongestPath(root.left,root.val)+GetmaxLongestPath(root.right,root.val);
        return Math.Max(maxLongestUnivaluePath,maxSumLongestPath);
    }
    
    public int GetmaxLongestPath(TreeNode root,int target){
        if(root==null || root.val!=target){
            return 0;
        }
        return 1+Math.Max(GetmaxLongestPath(root.left,target),GetmaxLongestPath(root.right,target));
    }
}
